File: canuck-trader-pro/backend/routes/questrade.py
"""
Questrade Routes
All /api/questrade/* endpoints for auth, accounts, candles, orders, paper trading, bot control.
"""

import asyncio
import time
import logging

from fastapi import APIRouter, HTTPException, Query, Request

router = APIRouter()
logger = logging.getLogger("questrade_routes")

# These will be set by app.py during startup
_questrade_service = None
_paper_trader = None
_stock_strategy = None
_bot_state = {
    "is_active": False,
    "is_paper": True,
    "account_id": None,
    "watchlist": ["SHOP", "TD", "RY", "BNS", "ENB", "CNR", "CP", "BMO", "BCE", "T"],
    "loop_task": None,
    "loop_ms": 5000,
}


def init_questrade(questrade_service, paper_trader, stock_strategy):
    global _questrade_service, _paper_trader, _stock_strategy
    _questrade_service = questrade_service
    _paper_trader = paper_trader
    _stock_strategy = stock_strategy


# ============================================
# AUTH
# ============================================

@router.post("/api/questrade/auth")
async def questrade_auth(request: Request):
    body = await request.json()
    refresh_token = body.get("refreshToken")
    is_practice = body.get("isPractice", False)
    try:
        _questrade_service.is_practice = is_practice
        await _questrade_service.authenticate(refresh_token)
        return {"success": True, "message": "Authenticated", "status": _questrade_service.get_status()}
    except Exception as e:
        raise HTTPException(401, str(e))


@router.get("/api/questrade/status")
async def questrade_status():
    return _questrade_service.get_status() if _questrade_service else {"authenticated": False}


# ============================================
# ACCOUNTS
# ============================================

@router.get("/api/questrade/accounts")
async def questrade_accounts():
    try:
        accounts = await _questrade_service.get_accounts()
        return {"accounts": accounts}
    except Exception as e:
        raise HTTPException(500, str(e))


@router.get("/api/questrade/balance/{account_id}")
async def questrade_balance(account_id: str):
    try:
        balance = await _questrade_service.get_balance(account_id)
        return balance
    except Exception as e:
        raise HTTPException(500, str(e))


@router.get("/api/questrade/positions/{account_id}")
async def questrade_positions(account_id: str):
    try:
        positions = await _questrade_service.get_positions(account_id)
        return {"positions": positions}
    except Exception as e:
        raise HTTPException(500, str(e))


# ============================================
# MARKET DATA
# ============================================

@router.get("/api/questrade/candles")
async def questrade_candles(
    ticker: str = Query(...),
    interval: str = Query("5m"),
    start: str | None = Query(None),
    end: str | None = Query(None),
):
    try:
        candles = await _questrade_service.get_candles_by_ticker(ticker, interval, start, end)
        return {"candles": candles, "count": len(candles)}
    except Exception as e:
        raise HTTPException(500, str(e))


@router.get("/api/questrade/search")
async def questrade_search(prefix: str = Query(...)):
    try:
        symbols = await _questrade_service.search_symbol(prefix)
        return {"symbols": symbols}
    except Exception as e:
        raise HTTPException(500, str(e))


@router.get("/api/questrade/symbols")
async def questrade_symbols(exchange: str = Query("TSX")):
    try:
        # Search popular Canadian tickers
        prefixes = ["SH", "TD", "RY", "BN", "CN", "CP", "ENB", "BMO", "BB", "AC"]
        all_symbols = []
        for prefix in prefixes:
            try:
                syms = await _questrade_service.search_symbol(prefix)
                filtered = [s for s in syms if s.get("listingExchange") == exchange and s.get("securityType") == "Stock"]
                all_symbols.extend(filtered)
            except Exception:
                pass
        # Dedupe
        seen = set()
        unique = []
        for s in all_symbols:
            if s["symbolId"] not in seen:
                seen.add(s["symbolId"])
                unique.append(s)
        return {"symbols": unique}
    except Exception as e:
        raise HTTPException(500, str(e))


# ============================================
# ORDERS
# ============================================

@router.post("/api/questrade/order")
async def questrade_order(request: Request):
    body = await request.json()
    account_id = body.get("accountId")
    if not account_id:
        raise HTTPException(400, "Missing accountId")
    try:
        result = await _questrade_service.place_order(account_id, body.get("order", {}))
        return result
    except Exception as e:
        raise HTTPException(500, str(e))


# ============================================
# PAPER TRADING
# ============================================

@router.get("/api/questrade/paper/summary")
async def paper_summary():
    return _paper_trader.get_account_summary() if _paper_trader else {}


@router.get("/api/questrade/paper/history")
async def paper_history(limit: int = Query(100)):
    return {"history": _paper_trader.get_history(limit) if _paper_trader else []}


@router.post("/api/questrade/paper/reset")
async def paper_reset(request: Request):
    body = await request.json()
    cash = body.get("startingCash")
    return _paper_trader.reset(cash) if _paper_trader else {"success": False}


# ============================================
# BOT CONTROL
# ============================================

async def _questrade_bot_loop():
    while _bot_state["is_active"]:
        try:
            for ticker in _bot_state["watchlist"]:
                try:
                    candles = await _questrade_service.get_candles_by_ticker(ticker, "5m")
                    if not candles or len(candles) < 30:
                        continue

                    signals = _stock_strategy.evaluate(candles, ticker)
                    consensus = _stock_strategy.get_consensus(signals)

                    if consensus["action"] == "BUY" and consensus["confidence"] >= 55:
                        price = candles[-1]["c"]
                        qty = max(1, int((_paper_trader.cash * 0.05) / price))
                        if qty > 0 and _paper_trader.cash >= qty * price:
                            _paper_trader.buy(ticker, qty, price)

                    elif consensus["action"] == "SELL" and ticker in _paper_trader.positions:
                        price = candles[-1]["c"]
                        qty = _paper_trader.positions[ticker]["quantity"]
                        _paper_trader.sell(ticker, qty, price)

                except Exception as e:
                    logger.error(f"Bot error for {ticker}: {e}")

            # Update prices
            for ticker in list(_paper_trader.positions.keys()):
                try:
                    candles = await _questrade_service.get_candles_by_ticker(ticker, "1m")
                    if candles:
                        _paper_trader.update_prices({ticker: candles[-1]["c"]})
                except Exception:
                    pass

        except Exception as e:
            logger.error(f"Questrade bot loop error: {e}")

        await asyncio.sleep(_bot_state["loop_ms"] / 1000)


@router.post("/api/questrade/bot/start")
async def start_questrade_bot(request: Request):
    if _bot_state["is_active"]:
        return {"success": False, "message": "Bot already running"}

    body = await request.json()
    _bot_state["is_active"] = True
    _bot_state["watchlist"] = body.get("watchlist", _bot_state["watchlist"])
    _bot_state["loop_task"] = asyncio.create_task(_questrade_bot_loop())
    return {"success": True, "message": "Questrade bot started"}


@router.post("/api/questrade/bot/stop")
async def stop_questrade_bot():
    _bot_state["is_active"] = False
    if _bot_state["loop_task"]:
        _bot_state["loop_task"].cancel()
        _bot_state["loop_task"] = None
    return {"success": True, "message": "Questrade bot stopped"}

File: routes/questrade.js
import { Router } from 'express';
import { createLogger } from '../services/logger.js';

const log = createLogger('Questrade');

export default function createQuestradeRouter(ctx) {
    const router = Router();
    const { questrade, paperTrader, strategyEngine, brain, questradeBotState, addLog } = ctx;

    // POST /questrade/auth
    router.post('/questrade/auth', async (req, res) => {
        try {
            const { refreshToken, isPractice } = req.body;
            if (refreshToken) {
                questrade.isPractice = isPractice ?? true;
            }
            await questrade.authenticate(refreshToken);
            res.status(200).json({ success: true, status: questrade.getStatus() });
        } catch (error) {
            res.status(400).json({ message: error.message });
        }
    });

    // GET /questrade/status
    router.get('/questrade/status', (req, res) => {
        res.status(200).json({
            questrade: questrade.getStatus(),
            bot: {
                isActive: questradeBotState.isActive,
                isPaper: questradeBotState.isPaper,
                watchlist: questradeBotState.watchlist,
            },
            paperTrading: {
                cash: paperTrader.portfolio.cash,
                positions: Object.keys(paperTrader.portfolio.positions).length,
                tradeCount: paperTrader.portfolio.history.length,
            }
        });
    });

    // GET /questrade/accounts
    router.get('/questrade/accounts', async (req, res) => {
        try {
            const accounts = await questrade.getAccounts();
            res.status(200).json({ accounts });
        } catch (error) {
            res.status(400).json({ message: error.message });
        }
    });

    // GET /questrade/balance/:accountId
    router.get('/questrade/balance/:accountId', async (req, res) => {
        try {
            const data = await questrade.getBalance(req.params.accountId);
            res.status(200).json(data);
        } catch (error) {
            res.status(400).json({ message: error.message });
        }
    });

    // GET /questrade/positions/:accountId
    router.get('/questrade/positions/:accountId', async (req, res) => {
        try {
            const positions = await questrade.getPositions(req.params.accountId);
            res.status(200).json({ positions });
        } catch (error) {
            res.status(400).json({ message: error.message });
        }
    });

    // GET /questrade/candles
    router.get('/questrade/candles', async (req, res) => {
        try {
            const { symbol, interval, start, end } = req.query;
            if (!symbol) return res.status(400).json({ message: 'symbol is required' });
            const candles = await questrade.getCandlesByTicker(symbol, interval || '1m', start, end);
            res.status(200).json({ candles });
        } catch (error) {
            res.status(400).json({ message: error.message });
        }
    });

    // GET /questrade/search
    router.get('/questrade/search', async (req, res) => {
        try {
            const { prefix } = req.query;
            if (!prefix) return res.status(400).json({ message: 'prefix is required' });
            const symbols = await questrade.searchSymbol(prefix);
            res.status(200).json({ symbols });
        } catch (error) {
            res.status(400).json({ message: error.message });
        }
    });

    // GET /questrade/symbols
    router.get('/questrade/symbols', async (req, res) => {
        try {
            const { exchange } = req.query;
            if (!exchange) return res.status(400).json({ message: 'exchange is required' });
            const symbols = await questrade.getSymbolsByExchange(exchange);
            res.status(200).json({ symbols });
        } catch (error) {
            res.status(400).json({ message: error.message });
        }
    });

    // POST /questrade/order
    router.post('/questrade/order', async (req, res) => {
        try {
            const { accountId, ticker, side, quantity, orderType, limitPrice } = req.body;
            if (!ticker || !side || !quantity) {
                return res.status(400).json({ message: 'ticker, side, and quantity are required' });
            }

            if (questradeBotState.isPaper) {
                const trade = await paperTrader.createOrder(ticker, side, quantity, orderType || 'MARKET', limitPrice);
                res.status(200).json({ success: true, trade, paper: true });
            } else {
                if (!accountId) return res.status(400).json({ message: 'accountId required for live trading' });
                const symbolId = await questrade.getSymbolId(ticker);
                const order = {
                    symbolId,
                    quantity,
                    icebergQuantity: quantity,
                    side: side === 'BUY' ? 'Buy' : 'Sell',
                    orderType: orderType === 'LIMIT' ? 'Limit' : 'Market',
                    timeInForce: 'Day',
                };
                if (limitPrice) order.limitPrice = limitPrice;
                const result = await questrade.placeOrder(accountId, order);
                res.status(200).json({ success: true, result, paper: false });
            }
        } catch (error) {
            res.status(400).json({ message: error.message });
        }
    });

    // GET /questrade/paper/summary
    router.get('/questrade/paper/summary', async (req, res) => {
        try {
            const summary = await paperTrader.getAccountSummary();
            res.status(200).json(summary);
        } catch (error) {
            res.status(400).json({ message: error.message });
        }
    });

    // GET /questrade/paper/history
    router.get('/questrade/paper/history', (req, res) => {
        res.status(200).json({ trades: paperTrader.getHistory() });
    });

    // POST /questrade/paper/reset
    router.post('/questrade/paper/reset', (req, res) => {
        const { balance } = req.body;
        paperTrader.reset(balance || 100000);
        res.status(200).json({ success: true, message: 'Paper trading reset' });
    });

    // POST /questrade/bot/start
    router.post('/questrade/bot/start', async (req, res) => {
        try {
            const { watchlist, isPaper, accountId } = req.body;
            if (questradeBotState.isActive) {
                return res.status(400).json({ message: 'Questrade bot already running' });
            }

            if (!questrade.isAuthenticated()) {
                await questrade.authenticate();
            }

            questradeBotState.isActive = true;
            questradeBotState.isPaper = isPaper !== false;
            questradeBotState.accountId = accountId || null;
            if (watchlist && Array.isArray(watchlist)) {
                questradeBotState.watchlist = watchlist;
            }

            questradeBotState.interval = setInterval(() => questradeBotLoop(), questradeBotState.loopMs);
            addLog(`[QUESTRADE BOT] Started (${questradeBotState.isPaper ? 'Paper' : 'Live'}) - Watchlist: ${questradeBotState.watchlist.join(', ')}`, 'INFO');
            res.status(200).json({ success: true, state: questradeBotState });
        } catch (error) {
            res.status(400).json({ message: error.message });
        }
    });

    // POST /questrade/bot/stop
    router.post('/questrade/bot/stop', (req, res) => {
        if (questradeBotState.interval) {
            clearInterval(questradeBotState.interval);
            questradeBotState.interval = null;
        }
        questradeBotState.isActive = false;
        addLog('[QUESTRADE BOT] Stopped', 'INFO');
        res.status(200).json({ success: true });
    });

    // Questrade Bot Loop (internal)
    function isMarketOpen() {
        const now = new Date();
        const etOffset = -5;
        const utcHour = now.getUTCHours();
        const utcMin = now.getUTCMinutes();
        const etHour = (utcHour + etOffset + 24) % 24;
        const etMinutes = etHour * 60 + utcMin;
        const dayOfWeek = now.getUTCDay();

        if (dayOfWeek === 0 || dayOfWeek === 6) return false;
        const marketOpen = 9 * 60 + 30;
        const marketClose = 16 * 60;
        return etMinutes >= marketOpen && etMinutes < marketClose;
    }

    async function questradeBotLoop() {
        if (!questradeBotState.isActive) return;

        try {
            if (!isMarketOpen()) {
                if (Math.random() < 0.01) {
                    addLog('[QUESTRADE BOT] Market closed - waiting', 'INFO');
                }
                return;
            }

            const watchlist = questradeBotState.watchlist;

            for (const ticker of watchlist) {
                try {
                    const candles = await questrade.getCandlesByTicker(ticker, '5m');
                    if (!candles || candles.length < 50) continue;

                    const signals = strategyEngine.evaluate(ticker, candles);
                    if (signals.length === 0) continue;

                    const bestSignal = signals.reduce((best, s) =>
                        s.confidence > best.confidence ? s : best, signals[0]
                    );

                    let aiDecision = { decision: 'YES', confidence: bestSignal.confidence * 100 };
                    if (bestSignal.confidence < 0.8) {
                        try {
                            const lastCandle = candles[candles.length - 1];
                            aiDecision = await brain.analyzeTradeOpportunity(
                                ticker, signals, {},
                                { price: lastCandle.c, volume: lastCandle.v }
                            );
                        } catch (e) { /* AI failure shouldn't block trades */ }
                    }

                    if (aiDecision.decision === 'NO') continue;

                    const lastPrice = candles[candles.length - 1].c;
                    const positionSize = questradeBotState.isPaper
                        ? Math.floor((paperTrader.portfolio.cash * 0.1) / lastPrice)
                        : 1;

                    if (positionSize <= 0) continue;

                    if (bestSignal.action === 'BUY') {
                        if (questradeBotState.isPaper) {
                            await paperTrader.createOrder(ticker, 'BUY', positionSize);
                        } else if (questradeBotState.accountId) {
                            const symbolId = await questrade.getSymbolId(ticker);
                            await questrade.placeOrder(questradeBotState.accountId, {
                                symbolId,
                                quantity: positionSize,
                                side: 'Buy',
                                orderType: 'Market',
                                timeInForce: 'Day',
                            });
                        }
                        addLog(`[QUESTRADE BOT] BUY ${positionSize} ${ticker} @ ${lastPrice} (${bestSignal.strategy}: ${bestSignal.reason})`, 'BUY');
                    } else if (bestSignal.action === 'SELL') {
                        const hasPosition = questradeBotState.isPaper
                            ? !!paperTrader.portfolio.positions[ticker]
                            : false;

                        if (hasPosition) {
                            const pos = paperTrader.portfolio.positions[ticker];
                            if (questradeBotState.isPaper) {
                                await paperTrader.createOrder(ticker, 'SELL', pos.quantity);
                            }
                            addLog(`[QUESTRADE BOT] SELL ${pos.quantity} ${ticker} @ ${lastPrice} (${bestSignal.strategy}: ${bestSignal.reason})`, 'SELL');
                        }
                    }

                    brain.reviewTrade({
                        ticker,
                        signal: bestSignal,
                        price: lastPrice,
                        timestamp: Date.now(),
                        paper: questradeBotState.isPaper,
                    });

                } catch (tickerError) {
                    if (Math.random() < 0.1) {
                        addLog(`[QUESTRADE BOT] Error on ${ticker}: ${tickerError.message}`, 'ERROR');
                    }
                }
            }
        } catch (error) {
            addLog(`[QUESTRADE BOT] Loop error: ${error.message}`, 'ERROR');
        }
    }

    return router;
}
